### PR TITLE
CREATE TABLE AS EXECUTE works.

### PR DESCRIPTION
**Pull request by guofengrichard**

### CODE REVIEW

File: src/backend/utils/cache/plancache.c [hunk 1]
```diff
@@ -542,18 +545,18 @@ RevalidateCachedQuery(CachedPlanSource *plansource, IntoClause *intoClause)
 	 */
 	if (intoClause)
 	{
-		CreateTableAsStmt *ctas_stmt;
+		SelectStmt *select;
```
Comment (yydzero) [inline src/backend/utils/cache/plancache.c:549]: why do we want to use SelectStmt instead of CreateTableAsStmt?

File: src/backend/cdb/cdbllize.c [hunk 2]
```diff
@@ -186,7 +186,7 @@ cdbparallelize(PlannerInfo *root,
 	switch (query->commandType)
 	{
 		case CMD_SELECT:
-			if (query->intoClause != NULL)
+			if (query->isCtas)
 			{
```
Comment (yydzero) [inline src/backend/cdb/cdbllize.c:190]: will this impact other queries which will use intoClause?

Comment (tvesely) [reply src/backend/cdb/cdbllize.c:190]: The intoClause went away in the 9.2 merge, but was put back in order to allow cdbparallelize to understand that it needs to send the tuples into a relation.

I believe that it also uses this code path to identify `SELECT INTO` commands, as they behave identically in upstream. In that case the `isCtas` variable name may be slightly misleading.

Comment (paul-guo-) [reply src/backend/cdb/cdbllize.c:190]: 'SELECT INTO' is transformed to CreateTableAsStmt earlier in transformTopLevelStmt().

isCtas seems to be a bit misleading. isForCtas, or just force_parallel to be more common?

File: src/backend/nodes/copyfuncs.c [hunk 3]
```diff
@@ -3134,6 +3134,7 @@ _copyQuery(const Query *from)
 	COPY_NODE_FIELD(constraintDeps);
 	COPY_NODE_FIELD(intoPolicy);
 	COPY_NODE_FIELD(intoClause);
+	COPY_SCALAR_FIELD(isCtas);
```
Comment (paul-guo-) [inline src/backend/nodes/copyfuncs.c:3137]: If you add a variable in Query, you need to add similar code in more than this file/functions, maybe equalfuncs.c, outfast.c, outfuncs.c, readfast.c, readfuncs.c, etc.

File: src/include/nodes/parsenodes.h [hunk 4]
```diff
@@ -189,6 +189,12 @@ typedef struct Query
 	 * and executor needs this info to specifiy correct target also.
 	 * */
 	IntoClause *intoClause;
+
+	/*
+	 * GPDB_92_MERGE_FIXME: Used by CTAS to indicate if this query is part of
+	 * CTAS statement.
+	 */
+	bool		isCtas;
```
Comment (paul-guo-) [inline src/include/nodes/parsenodes.h:197]: isCtas is not needed here since there is intoClause. Unless we really do dispatch CreateTableAsStmt, but that could be modified in another patch. In that case, intoclause is in CreateTableAsStmt so we do not keep another copy of intoClause here (delivered to plan from Query) for dispatcher and we could use a bool variable here of course.

File: src/include/nodes/parsenodes.h [hunk 5]
```diff
@@ -1180,6 +1186,13 @@ typedef struct SelectStmt
 	/* This field used by: SELECT INTO, CTAS */
 	/* GPDB_91_MERGE_FIXME: why is this not a DistributedBy*? */
 	Node *distributedBy;  /* GPDB: columns to distribute the data on. */
+
+	/*
+	 * GPDB_92_MERGE_FIXME: Used by CTAS.
```
Comment (paul-guo-) [inline src/include/nodes/parsenodes.h:1191]: I think we do not need GPDB_92_MERGE_FIXME in many places in this patch.

File: src/backend/utils/cache/plancache.c [hunk 6]
```diff
@@ -542,18 +545,18 @@ RevalidateCachedQuery(CachedPlanSource *plansource, IntoClause *intoClause)
...
 			ereport(ERROR,
 					(errcode(ERRCODE_WRONG_OBJECT_TYPE),
-					 errmsg("prepared statement is not a CreateTableAs Statement.")));
+					 errmsg("prepared statement is not a SELECT")));
 
-		ctas_stmt = (CreateTableAsStmt *) copyObject(plansource->raw_parse_tree);
+		select = (SelectStmt *) copyObject(plansource->raw_parse_tree);
 
-		ctas_stmt->into = copyObject(intoClause);
+		select->isCtas = true;
```
Comment (paul-guo-) [inline src/backend/utils/cache/plancache.c:557]: Maybe here we should set intoClause and isCtas is not needed for SelectStmt. Since for CTAS we are replanning, and we are replanning using the 'SELECT INTO' syntax.

File: src/backend/utils/cache/plancache.c (hunk 1 above)
Comment (paul-guo-) [reply src/backend/utils/cache/plancache.c:549]: Currently only Select is supported as the query for CTAS.

Comment (guofengrichard) [reply src/backend/utils/cache/plancache.c:549]: Hi @yydzero, what we want to re-plan at this place is the SelectStmt, rather than the CreateTableAsStmt.  Currently for CTAS, Postgres only supports the prepared statement to be a select statement. 
And from implementation view, if we use CreateTableAsStmt here, the result of the revalidation would be an UtilityStmt, and then we will encounter error in function `ExecuteQuery()`, because it is expecting the prepared statement to be a SELECT.
```
        pstmt = (PlannedStmt *) linitial(plan_list);
        if (!IsA(pstmt, PlannedStmt) ||
            pstmt->commandType != CMD_SELECT ||
            pstmt->utilityStmt != NULL)
            ereport(ERROR,
                    (errcode(ERRCODE_WRONG_OBJECT_TYPE),
                     errmsg("prepared statement is not a SELECT")));
```

File: src/backend/cdb/cdbllize.c (hunk 2 above)
Comment (guofengrichard) [reply src/backend/cdb/cdbllize.c:190]: @yydzero This would not impact other queries. `intoClause` has been removed by PG 9.2, we add it back here to correctly handle dispatch issue. Currently only CTAS (CREATE TABLE AS / SELECT INTO) statement would use `intoClause`, for other queries, it would be set to NULL.

Comment (guofengrichard) [reply src/backend/cdb/cdbllize.c:190]: @tvesely Yes. We need a more proper name here.

File: src/backend/utils/cache/plancache.c (hunk 1 above)
Comment (yydzero) [reply src/backend/utils/cache/plancache.c:549]: any possibility to make revalidation's result not an UtilityStmt?

or change pstmt->commandType != CMD_SELECT to 'pstmt->commandType != CMD_SELECT OR pstmt->commandType != CMD_UTILITY' for CTAS case?

does this make sense?

File: src/backend/utils/cache/plancache.c (hunk 6 above)
Comment (guofengrichard) [reply src/backend/utils/cache/plancache.c:557]: If `into` is set for the statement, the parser and analyzer would treat it as a `CreateTableAsStmt`, not a `SelectStmt` anymore. See function `transformTopLevelStmt()`.
And the result of the revalidation would be an UtilityStmt, not a PlannedStmt. The is not the expected result, and will error out in `ExecuteQuery()`.

File: src/include/nodes/parsenodes.h (hunk 4 above)
Comment (guofengrichard) [reply src/include/nodes/parsenodes.h:197]: If we retrieve the plan from plancache and do re-plan, there is no intoClause for query. That's why `isCtas` is needed here.

File: src/backend/utils/cache/plancache.c (hunk 1 above)
Comment (guofengrichard) [reply src/backend/utils/cache/plancache.c:549]: This may work but will diverge from upstream.  Currently PG doesn't support prepare a utility statement. So if we retrieve the plan from plancache and re-plan it, we are expecting it to be not an utility.

Comment (guofengrichard) [reply src/backend/utils/cache/plancache.c:549]: FYI. Using SelectStmt instead of CreateTableAsStmt here keeps the same behavior with master codes. Below is the code fragment from master `RevalidateCachedPlanWithParams()`:
```
        /*
         * If this is a CREATE TABLE AS, pass information about the
         * target table's distribution key to the planner.
         */
        if (intoClause)
        {
            SelectStmt *select;

            if (!IsA(plansource->raw_parse_tree, SelectStmt))
                ereport(ERROR,
                        (errcode(ERRCODE_WRONG_OBJECT_TYPE),
                         errmsg("prepared statement is not a SELECT")));

            select = (SelectStmt *) copyObject(plansource->raw_parse_tree);

            select->intoClause = copyObject(intoClause);

            rawtree = (Node *) select;
        }
```
